collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java

```diff
@@ -849,33 +809,38 @@ protected void deleteMediaFromContentProvider(Uri media) {
      * and if it's from Google Drive and not cached yet, we'll retrieve it using network.
      * This may take a long time.
      *
-     * @param selectedFile uri of the selected audio
+     * @param selectedFile uri of the selected media (audio/video/image/arbitrary file)
      */
     protected void saveChosenFile(Uri selectedFile) {
-        String extension = ContentResolverHelper.getFileExtensionFromUri(getContext(), selectedFile);
-        String instanceFolder = getFormController().getInstanceFile().getParent();
-        String destPath = instanceFolder + File.separator + System.currentTimeMillis() + extension;
-
-        try {
-            File chosenFile = MediaUtils.getFileFromUri(getContext(), selectedFile, MediaStore.Images.Media.DATA);
-            if (chosenFile != null) {
-                final File newFile = new File(destPath);
-                FileUtils.copyFile(chosenFile, newFile);
-                if (this instanceof BinaryWidget) {
-                    ((BinaryWidget) this).setBinaryData(newFile);
-                }
-                saveAnswersForCurrentScreen();
-            } else {
-                Timber.e("Could not receive chosen file");
-                ToastUtils.showShortToastInMiddle(R.string.error_occured);
-            }
-        } catch (GDriveConnectionException e) {
-            Timber.e("Could not receive chosen file due to connection problem");
-            ToastUtils.showLongToastInMiddle(R.string.gdrive_connection_exception);
-        }
-    }
 
-    private void refreshCurrentView() {
+        // show dialog
+        ((FormEntryActivity) getContext()).showDialog(SAVING_DIALOG);
+
+        ChosenFileSaver.createObservable(getContext(), selectedFile)
+                .doOnNext(saveResult -> {
+                    if (saveResult.isComplete() && this instanceof BinaryWidget) {
```

Comment:
So this has been the trickiest part and I spent so much time trying to figure out what was going wrong when I added the `Schedulers` to handle background threading. I haven't totally figure it out as yet but what I think is happening is there's some synchronous code that's not mixing well with the async approach we are using. One of the first things I noticed was that the `refreshCurrentView` wasn't being called after the operations were complete. So I added that to the `onSuccess` but that didn't work so some more investigations have to be done.

Reply:
Yes, so the main points I think we can follow to figure this out is 1) verify the data needed is being set to the widget and then check to see all the code that needs to execute once that's done so that the refreshing of the `ImageView` can display the picture.

Reply:
Yeah, I tried adding `refreshCurrentView` as well but it doesn't do anything. The value does not get saved so refreshing has no effect

Reply:
I am gonna work on some other stuff and then have another look at it with a fresh mind to see if I can see where exactly the issue is taking place.